### DISCUSSION TITLE
How to run tgcf as a service in background ?

### DISCUSSION THREAD

**Discussion by SHAliakbari** [aahnik/tgcf#219] [Q&A] [answered]
Hi, 
I clone the source and ran the app without problem and settings are OK.
now i want to run app in my Debian server as a service (systemctl).

how i could do this ?

**Comment by aahnik** [comment, marked as answer]
### Introduction

Tgcf has two modes. In past mode, the program exits when done. The live mode keeps running as long your server is up.

For more info read the [wiki page on Past vs Live ➥]( https://github.com/aahnik/tgcf/wiki/Past-vs-Live-modes-explained)

You can run the command
```shell
tgcf past
# ors
tgcf live
```

- For running tgcf live simply [deploy tgcf to Heroku](https://github.com/aahnik/tgcf/wiki/Deploy-to-Heroku) and it will run continuously.
- To run tgcf in past mode periodically (at an interval) use [GitHub Actions](https://github.com/aahnik/tgcf/wiki/Run-tgcf-in-past-mode-periodically).

### Run in background

There are **many ways** you can run tgcf continuously in the background.

- Using a terminal multiplexer like [`screen`](https://www.gnu.org/software/screen/)
- Running the docker container. Learn the [docker method ➥](https://github.com/aahnik/tgcf/wiki/Install-and-run-using-docker)
- Using [systemctl](https://www.geeksforgeeks.org/systemctl-in-unix/)

In the case of docker and systemctl, you can easily configure them to **restart the process when the server is rebooted**. Do some google searching for more info.

### Notes on `systemctl` service

When you properly install PostgreSQL or MongoDB, and such software, the systemctl services for those softwares are installed automatically.

But in the case of `tgcf`, **you need to manually create one**. **[See how ➥](https://github.com/aahnik/tgcf/wiki/Using-with-systemctl)**

This [article on Medium](https://abhinand05.medium.com/run-any-executable-as-systemd-service-in-linux-21298674f66f) will teach you how to create a **Linux service** to manage your programs.

To **see logs** from a running systemctl process, these StackOverflow posts ( [link](https://unix.stackexchange.com/questions/225401/how-to-see-full-log-from-systemctl-status-service) [link](https://unix.stackexchange.com/questions/475292/how-to-watch-output-from-systemd-service) ) are really helpful!

_The remaining part of this answer has been moved to a wiki page!_

**For a step by step, detailed guide for **using tgcf with systemctl** read [this wiki page ➥](https://github.com/aahnik/tgcf/wiki/Using-with-systemctl)**

**Reply by SHAliakbari** [reply]
Thanks . 
But when i use your syntax (path to .sh file) I get 203 error.
I used following code and everything works ok.

`ExecStart=/bin/sh '/usr/tgcf/tgcf/tgcf_start.sh'`

**Reply by aahnik** [reply]
You should not use my path. That's common sense. Use the path where the file is located in your server.

**Reply by SHAliakbari** [reply]
:-) yeah . I know . I just said i need to use `/bin/sh` to run `tgcf_start.sh` file

**Reply by aahnik** [reply]
Have you made the shell script executable ? In my case the default shell is bash, so in my shell script, the shebang was for bash.

In your server, it seems that sh is your default shell. So write the proper shebang in your shell script.

**Reply by SHAliakbari** [reply]
I got an error on time . 
Database is locked 

or something like that  after i ran it as service . 
after i reboot my os to prevent duplicate running tgcf  , it seems ok until now . 
in what situations such error may occurs ?

**Reply by aahnik** [reply]
When you are reporting an error, always show me your configuration and full logs. Use something like https://nekobin.com or https://gist.github.com

From my experience, I know what that means. 

You are right! That was caused by running multiple instances of tgcf (logged in with same session) at the same time.

Telethon uses a local database to cache certain things. When two instances run at same time and try to access the session files, that error is produced.

**Reply by SHAliakbari** [reply]
Thanks bro! you are amazing ;-)

**Reply by aahnik** [reply]
hi @SHAliakbari please star this repo, and share with your friends and social media. Please! Let us grow this project to 200 stars by this month.

**Reply by SHAliakbari** [reply]
Hey dear aahnik,
the service is running for tree days , from the day you help me . 
but i have one problem, some messages forwarder two times , duplicate . 
How could i get the log to check ? I ran it with -l command . in the systemd log i just got something like this : 

`New message received in -1001309370483   live.py:22` 

is it possible to see what text forwarded ?

**Reply by aahnik** [reply]
> but i have one problem, some messages forwarder two times , duplicate .

May be those messages were there two times, and later one was deleted in source. **It's impossible for tgcf to send a message twice.** YOU ARE MAKING SOME MISTAKE. Or may be, Telegram's servers, sent an update twice, accidentally in a rare case.

> I ran it with -l command 

what ? what do you mean ? what you ran ? speak details

I have already explained how to see logs for systemctl.

Read [this portion](https://github.com/aahnik/tgcf/wiki/Using-with-systemctl#running-via-systemctl) of the wiki. And in this answer https://github.com/aahnik/tgcf/discussions/219#discussioncomment-920558, I have already given links to two StackOverflow posts, that explain how to see full logs/live logs.

> is it possible to see what text forwarded ?

No, that is not logged.
